File: app/src/main/java/com/example/steaminvestmentmanager/MainActivity.java
package com.example.steaminvestmentmanager;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

//import com.fasterxml.jackson.core.JsonProcessingException;
//import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.*;
//import com.fasterxml.jackson.*;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;
import android.widget.PopupMenu;

import com.example.steaminvestmentmanager.utilclasses.*;
import java.util.ArrayList;
import java.util.Objects;

public class MainActivity extends AppCompatActivity {
    private static final ArrayList<SteamItem> steamItems = new ArrayList<>();

    private static final Gson gson = new Gson();

    public static SteamItemAdapter steamItemAdapter;
    private ListView steamItemsListView;
    private final MainActivity mainActivity = this;
    private final String preferenceName = "savedDataSIM";

    private final String TAG = "SteamIvan";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        steamItemsListView = findViewById(R.id.itemListView);
        steamItemAdapter = new SteamItemAdapter(getApplicationContext(), new ArrayList<>());
        setSteamItemsAdapter(steamItemAdapter);

        try {
            getSteamItemsFromPreference();
        }catch (Exception e) {
            Log.d(TAG, "onCreate: " + e);
        }

        ItemListUpdater.setMainActivity(this);
        ItemsUpdatingThread itemsUpdatingThread = new ItemsUpdatingThread();
        itemsUpdatingThread.start();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
    }

    @Override
    protected void onStop() {
        saveSteamItems();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    public void setSteamItemsAdapter(SteamItemAdapter steamItemAdapter) {
        runOnUiThread(() -> {
            try {
                steamItemsListView.setAdapter(steamItemAdapter);
            } catch (Exception ignored) {
            }
            steamItemsListView.setOnItemClickListener((parent, view, position, id) -> {
                SteamItem steamItem = Objects.requireNonNull(getSteamItems())[position];
                SteamItemInformationDialog steamItemInformationDialog = new SteamItemInformationDialog(steamItem);
                steamItemInformationDialog.show(getSupportFragmentManager(), null);
            });
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.user_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.open_menu) {
            showPopupMenu(findViewById(R.id.open_menu));
        }
        return true;
    }

    private void showPopupMenu(View v) {
        Context wrapper = new ContextThemeWrapper(getApplicationContext(), R.style.popupMenuStyle);
        PopupMenu popupMenu = new PopupMenu(wrapper, v);
        popupMenu.inflate(R.menu.user_popup_menu);
        popupMenu.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.add_steamItem) {
                EnteringURLDialog enteringURLDialog = new EnteringURLDialog(this);
                enteringURLDialog.show(getSupportFragmentManager(), null);
                return true;
            }
            if (item.getItemId() == R.id.open_settings) {
                SettingsDialog settingsDialog = new SettingsDialog();
                settingsDialog.show(getSupportFragmentManager(), null);
                return true;
            }
            return false;
        });
        popupMenu.show();
    }

    public void sendNewSteamItem(SteamItem addingSteamItem) {
        steamItems.add(addingSteamItem);
        runOnUiThread(() -> steamItemAdapter.add(addingSteamItem));
    }

    public static void deleteSelectedItem(SteamItem chosenSteamItem) {
        steamItems.remove(chosenSteamItem);
    }

    public static SteamItem[] getSteamItems() {
        SteamItem[] steamItemsArray;
        steamItemsArray = MainActivity.steamItems.toArray(new SteamItem[0]);
        return steamItemsArray;
    }

    private void clearSP() {
        getSharedPreferences(preferenceName, MODE_PRIVATE).edit().clear().apply();
    }

    private void getSteamItemsFromPreference() {
        SharedPreferences sharedPreferences = getSharedPreferences(preferenceName, MODE_PRIVATE);
        int steamItemLength = sharedPreferences.getInt("NUMBER_OF_STEAM_ITEMS", 0);
        if (steamItemLength != 0) {
            for (int i = 1; i <= steamItemLength; i++) {
                String jsonSteamItem = sharedPreferences.getString("item" + i, "");
                SteamItem currentSteamItem = gson.fromJson(jsonSteamItem, SteamItem.class);
                if (currentSteamItem != null && currentSteamItem.checkForBeingFull()) {
                    new Thread(currentSteamItem::updateIcon).start();
                    steamItems.add(currentSteamItem);
                    steamItemAdapter.add(currentSteamItem);
                }
            }
        }
        int userCurrency = sharedPreferences.getInt("USER_CURRENCY", 5);
        CurrencyData.setInitialCurrency(userCurrency);
    }

    private void saveSteamItems() {
        SharedPreferences sharedPreferences = getSharedPreferences(preferenceName, MODE_PRIVATE);
        SharedPreferences.Editor preferenceEditor = sharedPreferences.edit();
        preferenceEditor.clear();
        if (steamItems != null) {
            for (int i = 1; i <= steamItems.size(); i++) {
                steamItems.get(i - 1).setItemIcon(null);
                String itemJson = gson.toJson(steamItems.get(i - 1));
                preferenceEditor.putString("item" + i, itemJson);
            }
            preferenceEditor.putInt("NUMBER_OF_STEAM_ITEMS", steamItems.size());
        }
        preferenceEditor.putInt("USER_CURRENCY", CurrencyData.getCurrency());
        preferenceEditor.apply();
    }
}

File: app/src/main/java/com/example/steaminvestmentmanager/MainActivity.kt
package com.example.steaminvestmentmanager

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.ContextThemeWrapper
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import android.widget.PopupMenu
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.example.steaminvestmentmanager.utilclasses.CurrencyData
import com.example.steaminvestmentmanager.utilclasses.CurrencyExchangeDialog
import com.example.steaminvestmentmanager.utilclasses.EnteringURLDialog
import com.example.steaminvestmentmanager.utilclasses.ItemListUpdater
import com.example.steaminvestmentmanager.utilclasses.SettingsDialog
import com.example.steaminvestmentmanager.utilclasses.SteamItem
import com.example.steaminvestmentmanager.utilclasses.SteamItemAdapter
import com.example.steaminvestmentmanager.utilclasses.SteamItemInformationDialog
import com.google.gson.Gson
import java.util.Objects

class MainActivity : AppCompatActivity() {
    private var steamItemsListView: ListView? = null
    private val preferenceName = "savedDataSIM"
    private val TAG = "SteamIvan"
    private val itemsUpdatingThread = ItemsUpdatingThread()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(findViewById(R.id.toolbar))
        steamItemsListView = findViewById(R.id.itemListView)
        steamItemAdapter = SteamItemAdapter(applicationContext, ArrayList())
        setSteamItemsAdapter(steamItemAdapter)

        Thread {CurrencyData.updateExchangeRatios()}.start()
        for (i in CurrencyData.exchangeRatios.values) {
            Log.d("SteamIvan", "$i AAA")
        }

        try {
            retrieveItemsFromPreference()
        } catch (e: Exception) {
            Log.d(TAG, "onCreate: $e")
        }
        ItemListUpdater.mainActivity = this
        itemsUpdatingThread.start()
    }

    override fun onStop() {
        saveSteamItems()
        super.onStop()
    }

    override fun onResume() {
        Thread {CurrencyData.updateExchangeRatios()}.start()
        super.onResume()
    }

    private fun setSteamItemsAdapter(steamItemAdapter: SteamItemAdapter?) {
        runOnUiThread {
            try {
                steamItemsListView!!.adapter = steamItemAdapter
            } catch (ignored: Exception) { }
            steamItemsListView!!.onItemClickListener = OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                val steamItem = Objects.requireNonNull(getSteamItems())[position]
                val steamItemInformationDialog = SteamItemInformationDialog(steamItem)
                steamItemInformationDialog.show(supportFragmentManager, null)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.user_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.open_menu) {
            showPopupMenu(findViewById(R.id.open_menu))
        }
        return true
    }

    private fun showPopupMenu(v: View) {
        val wrapper: Context = ContextThemeWrapper(applicationContext, R.style.popupMenuStyle)
        val popupMenu = PopupMenu(wrapper, v)
        popupMenu.inflate(R.menu.user_popup_menu)
        popupMenu.setOnMenuItemClickListener { item: MenuItem ->
            if (item.itemId == R.id.add_steamItem) {
                val enteringURLDialog = EnteringURLDialog(this)
                enteringURLDialog.show(supportFragmentManager, null)
                return@setOnMenuItemClickListener true
            }
            if (item.itemId == R.id.open_settings) {
                val settingsDialog = SettingsDialog()
                settingsDialog.show(supportFragmentManager, null)
                return@setOnMenuItemClickListener true
            }
            if (item.itemId == R.id.currency_exchange_rate) {
                val currencyExchangeDialog = CurrencyExchangeDialog()
                currencyExchangeDialog.show(supportFragmentManager, null)
                return@setOnMenuItemClickListener true
            }
            if (item.itemId == R.id.scan_account_items) {
                itemsUpdatingThread.pauseThread()
                Thread {

                }
                return@setOnMenuItemClickListener true
            }
            false
        }
        popupMenu.show()
    }

    fun sendNewSteamItem(addingSteamItem: SteamItem) {
        steamItems.add(addingSteamItem)
        runOnUiThread { steamItemAdapter!!.add(addingSteamItem) }
    }

    private fun clearSP() {
        getSharedPreferences(preferenceName, MODE_PRIVATE).edit().clear().apply()
    }

    private fun retrieveItemsFromPreference() {
        val sharedPreferences = getSharedPreferences(preferenceName, MODE_PRIVATE)
        val steamItemLength = sharedPreferences.getInt("NUMBER_OF_STEAM_ITEMS", 0)
        if (steamItemLength != 0) {
            for (i in 1..steamItemLength) {
                val jsonSteamItem = sharedPreferences.getString("item$i", "")
                val currentSteamItem = gson.fromJson(jsonSteamItem, SteamItem::class.java)
                if (currentSteamItem != null && currentSteamItem.checkForBeingFull()) {
                    Thread { currentSteamItem.updateIcon(); }.start()
                    steamItems.add(currentSteamItem)
                    Log.d("SteamIvan", steamItems.toString())
                    steamItemAdapter!!.add(currentSteamItem)
                }
            }
        }
        val userCurrency = sharedPreferences.getInt("USER_CURRENCY", 5)
        CurrencyData.setInitialCurrency(userCurrency)
    }

    private fun saveSteamItems() {
        val sharedPreferences = getSharedPreferences(preferenceName, MODE_PRIVATE)
        val preferenceEditor = sharedPreferences.edit()
        preferenceEditor.clear()
        for (i in 1..steamItems.size) {
            steamItems[i - 1].itemIcon = null
            val itemJson = gson.toJson(steamItems[i - 1])
            preferenceEditor.putString("item$i", itemJson)
        }
        preferenceEditor.putInt("NUMBER_OF_STEAM_ITEMS", steamItems.size)
        preferenceEditor.putInt("USER_CURRENCY", CurrencyData.getCurrency())
        preferenceEditor.apply()
    }

    companion object {
        private val steamItems: ArrayList<SteamItem> = ArrayList()
        private val gson = Gson()
        @JvmField
        var steamItemAdapter: SteamItemAdapter? = null
        @JvmStatic
        fun deleteSelectedItem(chosenSteamItem: SteamItem) {
            steamItems.remove(chosenSteamItem)
        }

        @JvmStatic
        fun getSteamItems(): Array<SteamItem> {
            return steamItems.toTypedArray()
        }
    }
}